### ISSUE TITLE
BUG: ray 1.13 breaks CI, including test_binary and test_pickle, with circular import

### ISSUE THREAD

**Issue by mvashishtha** [modin-project/modin#4564]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS Monterey 12.2.1
- **Modin version** (`modin.__version__`): ba2b6ea47d81575e41af6c92ba4c2c3f8bcf54e2
- **Python version**: 3.9.12
- **Code we can use to reproduce**:

With ray 1.13, run

`pytest modin/pandas/test/dataframe/test_pickle.py`

and

`pytest modin/pandas/test/dataframe/test_binary.py` 

and maybe more.

Sample CI failure: https://github.com/modin-project/modin/runs/6819134820?check_suite_focus=true

### Describe the problem
I can't tell what's going wrong. The ray tasks raise a `RaySystemError` with stack trace for ` modin/pandas/test/dataframe/test_pickle.py::test_column_pickle[True]`:

<details>
<summary>Stack trace</summary>

```
ray.exceptions.RaySystemError: System error: cannot import name 'DataFrame' from partially initialized module 'pandas.core.frame' (most likely due to a circular import) (/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/frame.py)
traceback: Traceback (most recent call last):
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/ray/serialization.py", line 340, in deserialize_objects
    obj = self._deserialize_object(data, metadata, object_ref)
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/ray/serialization.py", line 237, in _deserialize_object
    return self._deserialize_msgpack_data(data, metadata_fields)
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/ray/serialization.py", line 192, in _deserialize_msgpack_data
    python_objects = self._deserialize_pickle5_data(pickle5_data)
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/ray/serialization.py", line 180, in _deserialize_pickle5_data
    obj = pickle.loads(in_band, buffers=buffers)
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/frame.py", line 157, in <module>
    from pandas.core.generic import NDFrame
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/generic.py", line 114, in <module>
    from pandas.core import (
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/indexing.py", line 45, in <module>
    import pandas.core.common as com
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/__init__.py", line 48, in <module>
    from pandas.core.api import (
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/api.py", line 48, in <module>
    from pandas.core.groupby import (
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/groupby/__init__.py", line 1, in <module>
    from pandas.core.groupby.generic import (
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/groupby/generic.py", line 70, in <module>
    from pandas.core.frame import DataFrame
ImportError: cannot import name 'DataFrame' from partially initialized module 'pandas.core.frame' (most likely due to a circular import) (/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/frame.py)
```
</details>

For `modin/pandas/test/dataframe/test_binary.py` for some reason I can't reproduce the failure on my laptop but here it is in CI: https://github.com/modin-project/modin/runs/6819134820?check_suite_focus=true

**Comment by devin-petersohn** [comment]
This has historically been a race condition between the import thread and the thread which starts running code. This code originally fixed that:

https://github.com/modin-project/modin/blob/ba2b6ea47d81575e41af6c92ba4c2c3f8bcf54e2/modin/core/execution/ray/common/utils.py#L226

@modin-project/modin-ray Do you know what changed with the import thread?

**Comment by vnlitvinov** [comment]
@devin-petersohn it seems we were getting lucky with that function, as it stopped doing what it was doing somewhere after 1.0: https://github.com/ray-project/ray/issues/12599

And they still don't have a good enough replacement based on https://github.com/ray-project/ray/issues/13908

😞

**Comment by mvashishtha** [comment]
We should pin ray < 1.13.0 because I don't see an easy fix. I'll send a fix for new issue #4566.

**Comment by devin-petersohn** [comment]
I am also testing adding a sleep during Ray initialization: https://github.com/modin-project/modin/pull/4568

**Comment by rkooo567** [comment]
I think we might have changed some parts around function import/export recently. Can you create an issue with Ray github? I can tag a guy who touched that part of codebase.

**Comment by devin-petersohn** [comment]
@rkooo567 Created issue here: https://github.com/ray-project/ray/issues/25675

**Comment by iycheng** [comment]
@devin-petersohn  we shouldn't use `ray.worker.global_worker.run_function_on_all_workers(_import_pandas)` since this is a private API.

Could you give more detail why it's used there?

Btw, ray also has modin tests in master which seems running good.

**Comment by devin-petersohn** [comment]
@iycheng we use that API because of the import thread race condition. This allowed all workers to import pandas before the execution of any tasks. The import thread race condition is the fundamental issue that we would hope to get fixed in Ray.

**Comment by iycheng** [comment]
@devin-petersohn could you share more details about which race condition? Like in what case it'll fail if we don't do this.

Btw, we updated the function importer to do lazy things right now:

- export if it doesn't exist in the global function table
- import if it doesn't exist locally

so we import as need compared with before where we import everything.

**Comment by devin-petersohn** [comment]
The race condition has historically been very difficult to reproduce regularly. At first, we were seeing it only on large CPU machines, but now it happens regularly in our testing infrastructure. 

Is there still a thread that imports libraries separately from deserialization and execution?

**Comment by iycheng** [comment]
For `run_function_on_all_workers`, yes, it's still working as before.

But for normal functions, we are not doing that any more since it doesn't make sense to import the function that's not used by the current worker.

**Comment by iycheng** [comment]
@devin-petersohn firstly, `run_function_on_all_workers` is not a public API and we shouldn't use this one outside of ray core. 

> The run_function_on_all_workers was intended to force all workers to fully import the pandas library before any execution starts.

Also this actually is incorrect, import thread is running async, it's not guaranteed and ray worker could be alive to take function before the function you want to run finished since all this happens in different thread (race condition)

**Comment by iycheng** [comment]
@devin-petersohn I tried it locally and https://github.com/modin-project/modin/pull/4594 at least can pass my local run with 

```
pytest modin/pandas/test/dataframe/test_pickle.py
pytest modin/pandas/test/dataframe/test_binary.py
```

The removal of `run_function_on_all_workers` is not necessary, but I just add it there (it needs to be removed for long term. all these are private APIs from ray).


I think the only thing I did is to copy `setup.cfg` to `modin/pandas/test/dataframe/setup.cfg`

I'm not quite sure about why it passes on my laptop.

I think you need to run the script with cwd as `modin/pandas/test/dataframe` otherwise, you'll need to put that dir into runtime env working dir.


Add main function into tests (example [here](https://gist.github.com/iycheng/3cdbe6cf70dd3c953c7e09286ba34ff4#file-gistfile1-txt-L86-L90)) and test with `python modin/pandas/test/dataframe/test_binary.py` will also work since the working dir will be changed to `modin/pandas/test/dataframe` directly in ray.

I'm not sure about the race condition you mentioned unless you can show me some script which can reproduce this easily.

**Comment by iycheng** [comment]
@devin-petersohn it seems the lint failed in that [PR](https://github.com/modin-project/modin/pull/4594) 
I don't have a setup dev env for Modin. Do you mind taking this over and see whether there is any issues left? Also this might not be the idea way to fix, but I think the issues is about the working dir for the test script.

**Comment by iycheng** [comment]
@mvashishtha are you running it with the PR I mentioned above? I didn't see any errors there. I'm running with master ray.

```
~/modin (fix-import) $ python test.py
UserWarning: Ray execution environment not yet initialized. Initializing...
To remove this warning, run the following python code before doing dataframe operations:

    import ray
    ray.init()

DeprecationWarning: `ray.ray_constants.MAC_DEGRADED_PERF_MMAP_SIZE_LIMIT` is a private attribute and access will be removed in a future Ray version.
UserWarning: Distributing <class 'numpy.ndarray'> object. This may take some time.
```

**Comment by mvashishtha** [comment]
@iycheng  no, but when I run it with your PR #4594 at commit 6702bae9e70c448fc698c9d150c69b54df040c68, the script in my comment no longer shows the bug.

@devin-petersohn do we need to do any further testing for the import race condition, or will #4594 work?

**Comment by vnlitvinov** [comment]
I wonder if this is related to https://bugs.python.org/issue38884, as I see in the [CI logs](https://github.com/modin-project/modin/runs/6819134820?check_suite_focus=true):
```
2022-06-09 19:25:08,443	WARNING worker.py:1404 -- Traceback (most recent call last):
  File "/usr/share/miniconda/envs/modin/lib/python3.8/site-packages/ray/_private/import_thread.py", line 193, in fetch_and_execute_function_to_run
    function({"worker": self.worker})
  File "/home/runner/work/modin/modin/modin/core/execution/ray/common/utils.py", line 94, in _import_pandas
    import pandas  # noqa F401
  File "/usr/share/miniconda/envs/modin/lib/python3.8/site-packages/pandas/__init__.py", line 48, in <module>
    from pandas.core.api import (
  File "/usr/share/miniconda/envs/modin/lib/python3.8/site-packages/pandas/core/api.py", line 48, in <module>
    from pandas.core.groupby import (
  File "/usr/share/miniconda/envs/modin/lib/python3.8/site-packages/pandas/core/groupby/__init__.py", line 1, in <module>
    from pandas.core.groupby.generic import (
  File "/usr/share/miniconda/envs/modin/lib/python3.8/site-packages/pandas/core/groupby/generic.py", line 70, in <module>
    from pandas.core.frame import DataFrame
  File "<frozen importlib._bootstrap>", line 988, in _find_and_load
  File "<frozen importlib._bootstrap>", line 149, in __enter__
  File "<frozen importlib._bootstrap>", line 94, in acquire
_frozen_importlib._DeadlockError: deadlock detected by _ModuleLock('pandas.core.frame') at 139708725689504
```

The more I read that Python bug the more I feel it is very likely the real root cause for what we have here...

**Comment by mvashishtha** [comment]
@vnlitvinov is there a case where the race condition persists even after #4594?

**Comment by iycheng** [comment]
@vnlitvinov  please see this [comment](https://github.com/modin-project/modin/issues/4564#issuecomment-1163774086) for details.
Keep `run_function_on_all_workers` also should just work. Two ways to make it work: 

- copy setup.cfg to the test dir (not sure why, but maybe this change the working dir to the place where the script is)
- run the test in the place where the script is

Btw, as I mentioned, this is [private API](https://github.com/ray-project/ray/blob/7a29461c40e5108a49e269af19515bef892e758a/python/ray/_private/worker.py#L680) in ray, please try not to use.

Please keep in mind **we might remove this function** since this is not used in ray core and it's async executed like before, which means you won't know when will this function be executed. It can be executed even after the function runs.

As the race condition mentioned in this ticket, if some reproducible scripts can be given, I'll be like to take a look.

**Comment by tianlinzx** [comment]
Any estimation when will Ray 1.13 be officially supported ?

**Comment by mvashishtha** [comment]
@tianlinzx the open PR #4603 linked to this issue should fix this bug. That PR is under active discussion. As soon as it's merged, you can [install Modin from the latest source](https://modin.readthedocs.io/en/stable/getting_started/installation.html#installing-from-the-github-master-branch), or wait for a Modin release that includes that fix.

**Comment by vnlitvinov** [comment]
@iycheng Modin no longer uses `run_function_on_all_workers` (will be released in 0.16 [soon™](https://wowwiki-archive.fandom.com/wiki/Soon)), so I guess Ray can continue deprecating this feature if we were the only library using it.

**Comment by tianlinzx** [comment]
When will modin 0.16 be released ?

**Comment by mvashishtha** [comment]
@tianlinzx I'm sorry for the late response. The Modin contributors are hoping to release 0.16 before the end of the next week (October 7).
